### PR TITLE
Fix for HighLevelWCS attribute error

### LINKED ISSUE #348: Failure w/astropy-dev from spectral-cube: APE14 issue?
I haven't tracked this all the way down yet, but this looks to me like a problem b/w reproject and astropy-dev?

```
   def test_mosaic_cubes(use_memmap, data_adv, use_dask, spectral_block_size):

        pytest.importorskip('reproject')

        # Read in data to use
        cube, data = cube_and_raw(data_adv, use_dask=use_dask)

        # cube is doppler-optical by default, which uses the rest wavelength,
        # which isn't auto-computed, resulting in nan pixels in the WCS transform
        cube._wcs.wcs.restwav = constants.c.to(u.m/u.s).value / cube.wcs.wcs.restfrq

        expected_wcs = WCS(combine_headers(cube.header, cube.header)).celestial

        # Make two overlapping cubes of the data
        part1 = cube[:, :round(cube.shape[1]*2./3.), :]
        part2 = cube[:, round(cube.shape[1]/3.):, :]

        assert part1.wcs.wcs.restwav != 0
        assert part2.wcs.wcs.restwav != 0

>       result = mosaic_cubes([part1, part2], order='nearest-neighbor',
                              roundtrip_coords=False,
                              spectral_block_size=spectral_block_size)

spectral_cube/tests/test_regrid.py:623:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
spectral_cube/cube_utils.py:832: in mosaic_cubes
    cube_repr = cube.reproject(header,
spectral_cube/utils.py:49: in wrapper
    return function(self, *args, **kwargs)
spectral_cube/spectral_cube.py:2697: in reproject
    newcube, newcube_valid = reproject_interp((data,
../astropy/astropy/utils/decorators.py:604: in wrapper
    return function(*args, **kwargs)
../python-reprojection/reproject/interpolation/high_level.py:121: in reproject_interp
    return _reproject_blocked(
../python-reprojection/reproject/utils.py:323: in _reproject_blocked
    da.store(
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/array/core.py:1236: in store
    compute_as_if_collection(Array, store_dsk, map_keys, **kwargs)
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/base.py:342: in compute_as_if_collection
    return schedule(dsk2, keys, **kwargs)
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:557: in get_sync
    return get_async(
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:500: in get_async
    for key, res_info, failed in queue_get(queue).result():
../../mambaforge/envs/py39forge/lib/python3.9/concurrent/futures/_base.py:439: in result
    return self.__get_result()
../../mambaforge/envs/py39forge/lib/python3.9/concurrent/futures/_base.py:391: in __get_result
    raise self._exception
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:542: in submit
    fut.set_result(fn(*args, **kwargs))
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:238: in batch_execute_tasks
    return [execute_task(*a) for a in it]
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:238: in <listcomp>
    return [execute_task(*a) for a in it]
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:229: in execute_task
    result = pack_exception(e, dumps)
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/local.py:224: in execute_task
    result = _execute_task(task, data)
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/optimization.py:990: in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/core.py:149: in get
    result = _execute_task(task, cache)
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/core.py:119: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../../mambaforge/envs/py39forge/lib/python3.9/site-packages/dask/array/core.py:523: in _pass_extra_kwargs
    return func(*args[len(keys) :], **kwargs)
../python-reprojection/reproject/utils.py:285: in reproject_single_block
    array, footprint = reproject_func(
../python-reprojection/reproject/interpolation/core.py:98: in _reproject_full
    _validate_wcs(wcs_in, wcs_out, array.shape, shape_out)
../python-reprojection/reproject/interpolation/core.py:28: in _validate_wcs
    if has_celestial(wcs_in) and not has_celestial(wcs_out):
../python-reprojection/reproject/wcs_utils.py:22: in has_celestial
    for world_axis_class in wcs.low_level_wcs.world_axis_object_classes.values():
../astropy/astropy/wcs/wcsapi/wrappers/sliced_wcs.py:295: in world_axis_object_classes
    keys_keep = [item[0] for item in self.world_axis_object_components]
../astropy/astropy/wcs/wcsapi/wrappers/sliced_wcs.py:291: in world_axis_object_components
    return [self._wcs.world_axis_object_components[idx] for idx in self._world_keep]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <iterator object at 0x17957bf40>

>   return [self._wcs.world_axis_object_components[idx] for idx in self._world_keep]
E   AttributeError: 'HighLevelWCSWrapper' object has no attribute 'world_axis_object_components'

../astropy/astropy/wcs/wcsapi/wrappers/sliced_wcs.py:291: AttributeError
```

### PR DESCRIPTION
**Pull request by keflavich**
This fixes #348, which is caused by a HighLevelWCSWrapper being passed where a LowLevelWCS object is expected.

@astrofrog I don't understand how this happens, as it's something created in a tall dask command tree by spectral-cube.  I can't come up with a minimal example that reproduces this error.

I added test coverage, but none of these tests reproduce the error.  They do, however, do add coverage by testing the `_reproject_blocked` code, which afaict was not covered at all (though I could easily have missed it)

### CODE REVIEW

**Comment by astrofrog** [comment]
I think the best fix might be to actually make it so that ``has_celestial`` can take a non-FITS WCS high level WCS object - basically we should check in the else clause that if it is a high level WCS object we should extract ``.low_level_wcs`` before proceeding with the check.

**Comment by keflavich** [comment]
I'm happy with that approach, I just wasn't sure it was intended.  I'd already prepped a version that worked like that; I'll push it shortly

**Comment by keflavich** [comment]
Ah, the weird thing is, what `has_celestial` is getting passed right now is something like `HighLevelWCS(SlicedWCS(HighLevelWCS)))`, which isn't really right

**Review by astrofrog — CHANGES_REQUESTED** [review]
Looks good overall but just a couple of comments!

File: reproject/utils.py [hunk 1]
```diff
@@ -277,7 +277,11 @@ def reproject_single_block(a, block_info=None):
         if a.ndim == 0 or block_info is None or block_info == []:
             return np.array([a, a])
         slices = [slice(*x) for x in block_info[None]["array-location"][-wcs_out.pixel_n_dim :]]
-        wcs_out_sub = HighLevelWCSWrapper(SlicedLowLevelWCS(wcs_out, slices=slices))
+
+        if isinstance(wcs_out, BaseHighLevelWCS):
+            wcs_out = SlicedLowLevelWCS(wcs_out.low_level_wcs, slices=slices)
```
Comment (astrofrog) [inline reproject/utils.py:282]: should just be ``wcs_out = wcs_out.low_level_wcs``

File: reproject/utils.py [hunk 2]
```diff
@@ -277,7 +277,11 @@ def reproject_single_block(a, block_info=None):
...
+        lowlevwcs = SlicedLowLevelWCS(wcs_out, slices=slices)
```
Comment (astrofrog) [inline reproject/utils.py:283]: Could you rename variable to ``low_level_wcs``?

**Comment by keflavich** [comment]
My spammed changes will need to be squashed of course

**Comment by keflavich** [comment]
@astrofrog I'm going to revert the change you suggested; it is invalid.  If we use `wcs_out = <something>`, it becomes a local variable in the scope of the function `reproject_single_block`, which isn't what we want.

File: reproject/tests/test_high_level.py [hunk 3]
```diff
@@ -117,8 +117,10 @@ def test_return_footprint(self, reproject_function):
 
 
 @pytest.mark.filterwarnings("ignore::FutureWarning")
-@pytest.mark.parametrize("projection_type, dtype", itertools.product(ALL_MODES, ALL_DTYPES))
-def test_surface_brightness(projection_type, dtype):
+@pytest.mark.parametrize(
+    "projection_type, dtype, block_size", itertools.product(ALL_MODES, ALL_DTYPES, (None, [2, 2]))
+)
+def test_surface_brightness(projection_type, dtype, block_size):
```
Comment (astrofrog) [inline reproject/tests/test_high_level.py:123]: Is this change still needed given the other tests which already check the blocked functionality?

**Review by astrofrog — COMMENTED** [review]
Looks good now! Just a small comment below, and otherwise just need to fix the code style and squash commits. Thanks!

File: reproject/tests/test_high_level.py (hunk 3 above)
Comment (keflavich) [reply reproject/tests/test_high_level.py]: nope.  I added a test that actually reproduces the error, though: the problem was that `reproject_interp` w/blocks was tested w/header but not w/WCS as output projection, but they behave differently

**Comment by keflavich** [comment]
One outstanding issue: the `test_has_celestial` code isn't actually testing what happens when it is given a HighLevelWCS with another HighLevelWCS as the `.low_level_wcs` attribute, because that can't happen - the initializers for HighLevelWCS don't allow it.  The issue was actually that a HighLevelWCS was being passed as an argument to a function that assumed only a LowLevelWCS was being passed.

So... I'm going to remove that unnecessary code.

**Review by Cadair — APPROVED** [review]
